ha/core/controllers/pcs/cluster_controller.py

```diff
@@ -121,13 +140,53 @@ def start(self) -> dict:
     @controller_error_handler
     def stop(self) -> dict:
         """
-        Stop cluster and all service.
+        Stop cluster and all service. It is Blocking call.
 
         Returns:
             ([dict]): Return dictionary. {"status": "", "msg":""}
                 status: Succeeded, Failed, InProgress
         """
-        raise HAUnimplemented("This operation is not implemented.")
+        status: str = ""
+        if not self._is_pcs_cluster_running():
+            raise ClusterManagerError("Cluster not running on current node."
+                        "To stop cluster, It should be running on current node.")
+        node_group: list = self._get_node_group()
+        local_node: str = ConfigManager.get_local_node()
+        Log.info(f"Node group for cluster start {node_group}, local node {local_node}")
+        self_group: list = list(filter(lambda group: (local_node in group), node_group))[0]
+        node_group.remove(self_group)
+        offline_nodes = self._get_filtered_nodes([NODE_STATUSES.POWEROFF.value])
+        # Stop cluster for other group
+        for node_subgroup in node_group:
+            for nodeid in node_subgroup:
+                if nodeid not in offline_nodes:
+                    if self.heal_resource(nodeid):
```

Comment:
Why is healing required while stopping the cluster? It should be taken care of in start.

Reply:
During cluster stop if some resource are in failed state then stop may take more time or can get block
to avoid running heal

It is useful mainly when resource are in block state on vm, on hw block handle by stonith

Reply:
then clearing fail count may put the service in blocked state again. I am not sure if the healing will cover all the scenarios. Why not put node in standby first?

Reply:
No same happen in standby case also.

Heal resource may solve problem.
But even if it fails again and got block, we are passing timeout to avoid it and we will get timeout error from pacemaker.